src/instrumentation.rs

```diff
@@ -0,0 +1,98 @@
+use crate::ffi;
+use crate::pyclass::boolean_struct::False;
+use crate::types::PyFrame;
+use crate::{Bound, PyAny, PyClass, PyObject, PyRefMut, PyResult, Python};
+use std::ffi::c_int;
+
+pub trait Event<'py>: Sized {
+    fn from_raw(what: c_int, arg: Option<Bound<'py, PyAny>>) -> PyResult<Self>;
+}
+
+pub enum ProfileEvent<'py> {
+    Call,
+    Return(Option<Bound<'py, PyAny>>),
+    CCall(Bound<'py, PyAny>),
+    CException(Bound<'py, PyAny>),
+    CReturn(Bound<'py, PyAny>),
+}
+
+impl<'py> Event<'py> for ProfileEvent<'py> {
+    fn from_raw(what: c_int, arg: Option<Bound<'py, PyAny>>) -> PyResult<ProfileEvent<'py>> {
+        let event = match what {
+            ffi::PyTrace_CALL => ProfileEvent::Call,
+            ffi::PyTrace_RETURN => ProfileEvent::Return(arg),
+            ffi::PyTrace_C_CALL => ProfileEvent::CCall(arg.unwrap()),
+            ffi::PyTrace_C_EXCEPTION => ProfileEvent::CException(arg.unwrap()),
+            ffi::PyTrace_C_RETURN => ProfileEvent::CReturn(arg.unwrap()),
+            _ => unreachable!(),
+        };
+        Ok(event)
+    }
+}
+
+pub trait Profiler: PyClass<Frozen = False> {
+    fn profile<'py>(
+        &mut self,
+        frame: Bound<'py, PyFrame>,
+        event: ProfileEvent<'py>,
+    ) -> PyResult<()>;
+}
+
+pub fn register_profiler<P: Profiler>(profiler: Bound<'_, P>) {
+    unsafe { ffi::PyEval_SetProfile(Some(profile_callback::<P>), profiler.into_ptr()) };
```

Comment:
Do you also want to support the new PyEval_SetProfileAllThreads function?

Also, there should be a way to clear tracing. Either with a separate function, or accepting an Option argument.

Finally, why needless change the name from `setprofile` / `SetProfile` to `register_profiler`?

Reply:
Yeah, I think supporting `PyEval_SetProfileAllThreads` makes sense. But I'll get the main bit working fully first.

The `register_profiler` naming was also inspired by the implementation in https://github.com/PyO3/pyo3/issues/4008#issuecomment-2027313230. I'd prefer to decide on final naming once the implementation is complete. `setprofile` could well make sense for symmetry.

Reply:
Looking again, I've decided to just rename now.

On the clearing tracing - I'm not sure there's an exposed way to do this in the Python C-API. We might need to wrap `sys.setprofile(None)` and `threading.setprofile(None)`. If I'm missing something here a pointer would be very welcome.